src/server.rs

```diff
@@ -137,9 +148,45 @@ async fn static_assets(req: Request<Body>) -> (StatusCode, HeaderMap, Body) {
                 return (StatusCode::INTERNAL_SERVER_ERROR, headers, body);
             }
         };
-        let script = format!(include_str!("templates/websocket.html"), addr);
+        let script = format_script(addr, is_reload, false);
         file = format!("{text}{script}").into_bytes();
+    } else if !HARD.get().copied().unwrap_or(false) {
+        // allow client to cache assets for a smoother reload.
+        // client handles preloading to refresh cache before reloading.
+        headers.append(
+            header::CACHE_CONTROL,
+            HeaderValue::from_str("max-age=30").unwrap(),
+        );
     }
 
     (StatusCode::OK, headers, Body::from(file))
 }
+
+/// JS script containg a function that takes in the address and connects to the websocket.
+const WEBSOCKET_FUNCTION: &str = include_str!("templates/websocket.js");
+
+/// JS script to inject to the HTML on reload so the client
+/// knows it's a successful reload.
+const RELOAD_PAYLOAD: &str = include_str!("templates/reload.js");
+
+/// Inject the address into the websocket script and wrap it in a script tag
+fn format_script(addr: &str, is_reload: bool, is_error: bool) -> String {
+    match (is_reload, is_error) {
+        // successful reload, inject the reload payload
+        (true, false) => format!("<script>{}</script>", RELOAD_PAYLOAD),
+        // failed reload, don't inject anything so the client polls again
+        (true, true) => String::new(),
```

Comment:
Why distinguish between these two situations by `is_error`? Currently when accessing a page that does not exist, the 404 page will not keep reloading, instead, it will only keep reloading after some other files change, which does not make sense.

Reply:
The `?reload` parameter is for the client to probe if the page is ready to reload. If it is, the server's response will include an injected meta tag. If it's not, the server will send the 404 page instead without the meta tag. The third arm is for normal loading and we do inject the websocket client to error page so it can reload.

We need the meta tag because iframes only send onload event, even when the status is 404